### PR TITLE
shell: implement Shell::is_runnable and highlight commands

### PR DESCRIPTION
**Pull request by math2001**
Hey Serenity!

Inspired from the fish shell, the terminal now highlights the current command based on whether it exists or not (red if it is invalid, and bold white if it's valid).

serenity-terminal-command-highlight

I add the method `is_runnable` to the shell. I'm not sure if that's the best name/go to way about it. Feedback more than welcome :slightly_smiling_face: 

Also, the builtin `alias` wasn't refreshing the cache `cached_path` when an aliases was added. That meant that no aliases were cached because the cache is populated *before* the rc files are loaded. So this is fixed (each alias call insert the alias in the cache).

Shouldn't `cached_path` be called `m_cached_path`? Or am i missing something? 

Thanks this awesome project, it's really inspiring :heart:

### CODE REVIEW

**Comment by awesomekling** [comment]
cc @alimpfard

File: Shell/AST.cpp [hunk 1]
```diff
@@ -367,9 +367,15 @@ RefPtr<Value> BarewordLiteral::run(RefPtr<Shell>)
 void BarewordLiteral::highlight_in_editor(Line::Editor& editor, Shell& shell, HighlightMetadata metadata)
 {
     if (metadata.is_first_in_list) {
-        editor.stylize({ m_position.start_offset, m_position.end_offset }, { Line::Style::Bold });
+        if (shell.is_runnable(m_text))
+            editor.stylize({ m_position.start_offset, m_position.end_offset }, { Line::Style::Bold });
+        else {
```
Comment (alimpfard) [inline Shell/AST.cpp:372]: Please either omit the braces here, or add one to the `if` above.

File: Shell/Builtin.cpp [hunk 2]
```diff
@@ -64,6 +64,20 @@ int Shell::builtin_alias(int argc, const char** argv)
...
+                },
+                &index);
+
+            if (match)
+                continue;
+
+            while (strcmp(cached_path[index].characters(), parts[0].characters()) < 0) {
+                index++;
+            }
+            cached_path.insert(index, parts[0]);
```
Comment (alimpfard) [inline Shell/Builtin.cpp:80]: How about a separate function for this?

File: Shell/AST.cpp [hunk 3]
```diff
@@ -367,9 +367,15 @@ RefPtr<Value> BarewordLiteral::run(RefPtr<Shell>)
...
+            editor.stylize({ m_position.start_offset, m_position.end_offset }, { Line::Style::Foreground(Line::Style::XtermColor::Red) });
```
Comment (alimpfard) [inline Shell/AST.cpp:373]: Let's keep it bold either way, it's nice to have an indicator that doesn't depend on the text itself.

Also, I'm not too sure about the color here, red is also used to highlight syntax errors...but no other good alternative comes to mind, feel free to disregard this bit.

**Comment by alimpfard** [comment]
> Shouldn't cached_path be called m_cached_path?

It should, but no one has gotten around to dealing with it yet (same applies for most of the state in the Shell).
it should be simple to fix now that Shell's behavior is part of the class itself.

File: Shell/Builtin.cpp (hunk 2 above)
Comment (math2001) [reply Shell/Builtin.cpp:80]: Oh yeah I'll do this

File: Shell/AST.cpp (hunk 1 above)
Comment (math2001) [reply Shell/AST.cpp:372]: Oops that wasn't intended

File: Shell/AST.cpp (hunk 3 above)
Comment (math2001) [reply Shell/AST.cpp:373]: Bold red seems a bit in your face IMO. Are you sure? I'll push it if you want though.

serenity-terminal

Comment (alimpfard) [reply Shell/AST.cpp:373]: Well...the non-bold variant looks better, I guess.
let's leave it non-bold then.

**Pull request merged**